### PR TITLE
#167190431 Implement JWT authentication

### PR DESCRIPTION
**Pull request by chuxmykel**
#### What does this PR do?
- This PR implements user authentication.

#### Description of the task to be completed
- implement sign-in endpoint
- implement sign up endpoint
- provide users with JWT on sign up or sign in

#### How can this be manually tested?
- visit the staging link that will be generated from this pull request and hit the POST `/api/v1/users/login` route with an `application/json` payload in the format: 
```
{
  "user": {
    "email": "test@email.com",
    "password": "P@5SW01D",
  }
}
```

or

- visit the staging link that will be generated from this pull request and hit the POST `/api/v1/users` route with an `application/json` payload in the format: 

```
{
  "user": {
    "email": "test@email.com",
    "password": "P@5SW01D",
    "firstname": "peter",
    "lastname": "martin",
    "username": "pmartin",
  }
}
```
#### Any background context you want to provide?
- N/A

#### What are the relevant Pivotal Tracker stories? 
- [#167190431](https://www.pivotaltracker.com/story/show/167190431)

### CODE REVIEW

File: server/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      id: user.id, firstname, lastname, username, email,
+    });
+    return utils.successStat(res, 201, 'user', {
+      id: user.id, token, username, firstname, lastname, email,
+    });
+  }
+
+  /**
+    * @static
+    * @description Allows a user to sign up
```
Comment (cvjude) [inline server/controllers/userController.js]: This should be sign in Chux

File: server/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    const token = auth.generateToken({
+      id: user.id,
+      firstname: user.firstname,
+      lastname: user.firstname,
+      username: user.username,
+      email,
+      bio: user.bio,
+      image: user.image,
+    });
```
Comment (cvjude) [inline server/controllers/userController.js]: I think we should just have the users id and maybe the username in the token, I feel the rest of those details may not be needed

File: server/controllers/userController.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    * @description Allows a user to sign up
+    * @param {Object} req - Request object
+    * @param {Object} res - Response object
+    * @returns {Object} object containing user data and access Token
+    * @memberof UserController
+    */
+  static async signUp(req, res) {
+    const {
+      firstname, lastname, email, username, password
+    } = req.body.user;
```
Comment (cvjude) [inline server/controllers/userController.js]: Chux why do we have req.body.user here and in signin

File: server/tests/testData/user.data.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
+const user = [
+  {
+    user: {
```
Comment (cvjude) [inline server/tests/testData/user.data.js]: Here too, this should just be coming from the body

File: server/controllers/userController.js (hunk 2 above)
Comment (chuxmykel) [reply server/controllers/userController.js]: I was trying to cover for the front end in my thought process. Considering we will be working with react. Maybe we can link up for quick sync with the team so we can talk about it in detail.

**Review by chiboycalix — APPROVED** [review]
Aside Jude's Comment, everything looks good to me

File: app.json [hunk 5]
```diff
@@ -2,9 +2,12 @@
...
+  "env": {
+    "DATABASE_URL": "postgres://unfmehko:931GEzpHkZL-5HGjnjmxRhMXI5KwNVea@raja.db.elephantsql.com:5432/unfmehko",
+    "SECRET_KEY": "judeAndDominicAreFriendsWithKafilat"
+  },
   "formation": {},
   "addons": [],
   "buildpacks": [],
   "stack": "heroku-18"
-}
+}
```
Comment (topseySuave) [inline app.json]: Can you fix this newlIne error

File: server/controllers/userController.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      where: {
+        [Op.or]: [{ email }, { username }]
+      }
+    });
+    if (existingUser) {
+      return utils.errorStat(res, 409, 'User Already Exists');
+    }
+    const newUser = { ...req.body.user, password: hash.hashPassword(password) };
+    const user = await models.User.create(newUser);
+    const token = auth.generateToken({ id: user.id, username });
```
Comment (topseySuave) [inline server/controllers/userController.js]: Why is the token only generated with only the user's id and username what about the email for proper validation?

File: server/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    const { email, password } = req.body.user;
+    const user = await models.User.findOne({ where: { email } });
+
+    if (!user) return utils.errorStat(res, 401, 'Incorrect Login information');
+    const matchPasswords = hash.comparePassword(password, user.password);
+    if (!matchPasswords) return utils.errorStat(res, 401, 'Incorrect Login information');
+
+    return utils.successStat(res, 200, 'user', {
+      id: user.id,
+      token: await auth.generateToken({ id: user.id, username: user.username }),
```
Comment (topseySuave) [inline server/controllers/userController.js]: Same here for the email assigned to the token

File: server/controllers/userController.js (hunk 6 above)
Comment (chuxmykel) [reply server/controllers/userController.js]: I'll fix this now.

File: app.json (hunk 5 above)
Comment (chuxmykel) [reply app.json]: Sure. I will do that now.

Comment (chuxmykel) [reply app.json]: Fixed

File: server/controllers/userController.js (hunk 7 above)
Comment (chuxmykel) [reply server/controllers/userController.js]: Fixed

File: server/controllers/userController.js (hunk 6 above)
Comment (chuxmykel) [reply server/controllers/userController.js]: Fixed

**Review by Hadeneekeh — APPROVED** [review]

**Review by obayomi96 — APPROVED** [review]
Good job on the authentication implementation @chuxmykel and @cvjude.

**Review by cvjude — APPROVED** [review]
LGTM

**Review by topseySuave — APPROVED** [review]

**Review by encodedBicoding — APPROVED** [review]

**Pull request merged**